src/components/select/select-list/select-list.component.js

```diff
@@ -0,0 +1,170 @@
+
+      onSelect({ text, value, selectionType: 'enterKey' });
+    } else if (isNavigationKey(event)) {
+      const nextIndex = getNextIndexByKey(event.which, currentOptionsListIndex, lastIndex);
+
+      highlightItem(nextIndex, optionList);
+    }
+  }, [currentOptionsListIndex, children, onSelectListClose, onSelect, highlightItem]);
+
+  const repositionList = useCallback(() => {
+    if (anchorElement) {
+      const inputBoundingRect = anchorElement.getBoundingClientRect();
+      const top = `${inputBoundingRect.top + inputBoundingRect.height + window.pageYOffset}px`;
+      const width = `${inputBoundingRect.width}px`;
+      const left = `${inputBoundingRect.left}px`;
+
+      listRef.current.setAttribute('style', `top: ${top}; width: ${width}; left: ${left}`);
+    }
+  }, [anchorElement, listRef]);
+
+  const getIndexOfMatch = useCallback((valueToMatch) => {
+    return React.Children.toArray(children).findIndex(child => child.props.value === valueToMatch);
+  }, [children]);
+
+  useEffect(() => {
+    const keyboardEvent = 'keydown';
+
+    document.addEventListener(keyboardEvent, handleGlobalKeydown);
+
+    return function cleanup() {
+      document.removeEventListener(keyboardEvent, handleGlobalKeydown);
+    };
+  }, [handleGlobalKeydown]);
+
+  useEffect(() => {
+    if (!filterText) {
+      return;
+    }
+
+    setCurrentOptionsListIndex((previousIndex) => {
+      const match = getNextChildByText(filterText, children, previousIndex);
+
+      if (!match) {
+        return previousIndex;
+      }
+
+      const indexOfMatch = getIndexOfMatch(match.props.value);
+
+      updateListScrollTop(indexOfMatch, listRef.current);
+
+      return indexOfMatch;
+    });
+  }, [children, filterText, getIndexOfMatch, listRef]);
+
+  useEffect(() => {
+    repositionList();
+  }, [repositionList]);
+
+  useEffect(() => {
+    if (!highlightedValue) {
+      return;
+    }
+
+    const indexOfMatch = getIndexOfMatch(highlightedValue);
+
+    setCurrentOptionsListIndex(indexOfMatch);
+    updateListScrollTop(indexOfMatch, listRef.current);
+  }, [getIndexOfMatch, highlightedValue, listRef]);
+
+  function getChildrenWithListProps() {
+    return React.Children.map(children, (child, index) => {
+      const newProps = {
+        onSelect: handleSelect,
+        index,
+        selectedIndex: currentOptionsListIndex
+      };
+
+      return React.cloneElement(child, newProps);
+    });
+  }
+
+  function handleSelect(optionData) {
+    onSelect({ ...optionData, selectionType: 'click' });
+  }
+
+  function isNavigationKey(keyEvent) {
+    return Events.isDownKey(keyEvent) || Events.isUpKey(keyEvent)
+    || Events.isHomeKey(keyEvent) || Events.isEndKey(keyEvent);
+  }
+
+  return (
+    <Portal onReposition={ repositionList }>
+      <StyledSelectList
+        id={ id }
+        data-element='select-list'
+        role='listbox'
+        ref={ listRef }
+      >
+        { getChildrenWithListProps() }
```

Comment:
We're currently cloning the children every time we render, do we need to?

Reply:
I think that we cannot memoize this, so we cannot do it better